### PR TITLE
Support multiple assessments per trace in MemAlign optimizer

### PR DESCRIPTION
**Pull request by veronicalyu320**
### Related Issues/PRs

Relates to MemAlign optimizer feature - improving multi-assessment handling

### What changes are proposed in this pull request?

Modified `trace_to_dspy_example()` in the MemAlign optimizer to properly handle multiple human assessments with the same judge name on a single trace:

- **No conflict**: When all assessments agree on the label, returns an example for each assessment (stronger training signal)
- **Conflict with majority**: Returns examples only for the majority label, logs warning with discarded assessment details
- **Tie**: Returns examples for the group with the most recent assessment timestamp

Key changes:
- Changed `trace_to_dspy_example()` return type from `Optional[dspy.Example]` to `list[dspy.Example]`
- Added `_resolve_assessment_conflicts()` for majority voting with recency-based tie-breaking
- Added `_format_assessment_details()` and `_get_assessment_timestamp()` helper functions
- Updated `optimizer.py` callers to handle list return type
- Warning messages include `assessment_id`, `source_id`, and `create_time_ms` for debugging

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests

Added parametrized tests in `test_dspy_utils.py` covering:
- Multi-assessment with no conflict (1/2/3 agreeing assessments)
- Majority voting (3v1, 2v1 scenarios)
- Tie-breaking by recency (2v2, 3-way ties)
- Warning message content validation
- Edge cases (no timestamps, LLM-only, mixed sources, wrong judge name)
- Optimizer integration

All 73 tests pass locally.

### Does this PR require documentation update?

- [x] No.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

The MemAlign optimizer now handles multiple human assessments per trace by using majority voting. When assessments conflict, it uses the majority label (or most recent on ties) and logs a warning with details of discarded assessments.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 1]
```diff
@@ -335,10 +336,111 @@ def convert_litellm_to_mlflow_uri(litellm_model: str) -> str:
         raise MlflowException(f"Failed to convert LiteLLM format to MLflow URI: {e}")
 
 
-def trace_to_dspy_example(trace: Trace, judge: Judge) -> Optional["dspy.Example"]:
+def _format_assessment_details(assessment) -> str:
+    """Format assessment details for warning messages."""
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: nit: let's remove 1 line docstrings.

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 2]
```diff
@@ -335,10 +336,111 @@ def convert_litellm_to_mlflow_uri(litellm_model: str) -> str:
...
+    if feedback := getattr(assessment, "feedback", None):
+        parts.append(f"label='{feedback.value}'")
+    if source := getattr(assessment, "source", None):
+        parts.append(f"source_id='{source.source_id}'")
+    if create_time_ms := getattr(assessment, "create_time_ms", None):
+        parts.append(f"create_time_ms={create_time_ms}")
+    return ", ".join(parts)
+
+
+def _get_assessment_timestamp(assessment) -> int:
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: does this need to be a helper since it's only used once and it's one line?

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 3]
```diff
@@ -335,10 +336,111 @@ def convert_litellm_to_mlflow_uri(litellm_model: str) -> str:
         raise MlflowException(f"Failed to convert LiteLLM format to MLflow URI: {e}")
 
 
-def trace_to_dspy_example(trace: Trace, judge: Judge) -> Optional["dspy.Example"]:
+def _format_assessment_details(assessment) -> str:
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: let's add type hints here

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 4]
```diff
@@ -335,10 +336,111 @@ def convert_litellm_to_mlflow_uri(litellm_model: str) -> str:
...
+    """Format assessment details for warning messages."""
+    parts = []
+    if assessment_id := getattr(assessment, "assessment_id", None):
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: I think we should just fail if some of these attributes are not available - we can expect if they're not present, something is wrong. specifically assessment id, label, and create time.

Comment (smoorjani) [reply mlflow/genai/judges/optimizers/dspy_utils.py]: also, why do we need to pass assessment ID?

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 5]
```diff
@@ -368,55 +470,44 @@ def trace_to_dspy_example(trace: Trace, judge: Judge) -> Optional["dspy.Example"
...
             _logger.warning(
                 f"No human assessment found for judge '{judge.name}' in trace {trace.info.trace_id}"
             )
-            return None
+            return []
 
-        if not expected_result.feedback:
-            _logger.warning(f"No feedback found in assessment for trace {trace.info.trace_id}")
-            return None
+        # Filter out assessments without feedback
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: nit: probably don't need these one-liner comments, code is self-explanatory

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 6]
```diff
@@ -335,10 +336,111 @@ def convert_litellm_to_mlflow_uri(litellm_model: str) -> str:
...
+        parts.append(f"create_time_ms={create_time_ms}")
+    return ", ".join(parts)
+
+
+def _get_assessment_timestamp(assessment) -> int:
+    """Get the timestamp of an assessment, defaulting to 0 if not available."""
+    return getattr(assessment, "create_time_ms", None) or 0
+
+
+def _log_conflict_warning(
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: this signature is a bit complex - I'd rather we do logging in line rather than have a separate helper for this

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 7]
```diff
@@ -335,10 +336,111 @@ def convert_litellm_to_mlflow_uri(litellm_model: str) -> str:
...
+
+def _resolve_assessment_conflicts(
+    assessments: list, trace_id: str, judge_name: str
+) -> list:
+    """
+    Resolve conflicts when multiple assessments have different labels.
+
+    Args:
+        assessments: List of matching human assessments (already filtered by judge name)
+        trace_id: Trace ID for logging
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: these two (`trace_id` and `judge_name` are only passed for logging) - can we do this one-level up? maybe we can check if the length of assessments passed in is not equal to the length returned and if so, log the trace id/judge name

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 8]
```diff
@@ -335,10 +336,111 @@ def convert_litellm_to_mlflow_uri(litellm_model: str) -> str:
...
+        judge_name: Judge name for logging
+
+    Returns:
+        List of assessments to use (majority label, or most recent group on tie)
+    """
+    if not assessments:
+        return []
+
+    # Group assessments by their feedback value (label)
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: nit: same as before on one-liner comments

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 9]
```diff
@@ -335,10 +336,111 @@ def convert_litellm_to_mlflow_uri(litellm_model: str) -> str:
...
+        resolution = f"Using majority label '{winning_label}'"
+
+    _logger.warning(
+        f"{prefix}: {label_counts}. {resolution}. "
+        f"Discarded {len(discarded)} assessment(s):\n  - {discarded_details}"
+    )
+
+
+def _resolve_assessment_conflicts(
+    assessments: list, trace_id: str, judge_name: str
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]:
```suggestion
    assessments: list[Assessment], trace_id: str, judge_name: str
```

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 10]
```diff
@@ -335,10 +336,111 @@ def convert_litellm_to_mlflow_uri(litellm_model: str) -> str:
...
+) -> list:
+    """
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: probably don't need docstring for internal methods

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 11]
```diff
@@ -335,10 +336,111 @@ def convert_litellm_to_mlflow_uri(litellm_model: str) -> str:
...
+
+    is_tie = len(majority_labels) > 1
+    if is_tie:
+        # Tie - use the group with the most recent assessment
+        def get_group_max_timestamp(label: str) -> int:
+            return max(_get_assessment_timestamp(a) for a in groups[label])
+
+        winning_label = max(majority_labels, key=get_group_max_timestamp)
+    else:
+        winning_label = majority_labels[0]
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: does this work?

```suggestion
 def group_priority(label: str) -> tuple[int, int]:                         
      group = groups[label]                                                  
      return len(group), max(_get_assessment_timestamp(a) for a in group)    
                                                                             
  winning_label = max(groups, key=group_priority)  
```

File: tests/genai/judges/optimizers/test_dspy_utils.py [hunk 12]
```diff
@@ -469,3 +540,263 @@ class NonDictDemo:
...
+@pytest.mark.parametrize(
+    ("num_assessments", "expected_label"),
+    [
+        (1, "pass"),  # Single assessment returns single-item list
+        (2, "pass"),  # Two agreeing assessments return both
+        (3, "fail"),  # Three agreeing assessments return all
+    ],
+)
+def test_trace_to_dspy_example_multi_assessment_no_conflict(mock_judge, num_assessments, expected_label):
+    dspy = pytest.importorskip("dspy", reason="DSPy not installed")
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/test_dspy_utils.py]: we should assume dspy is installed - if it's not, something is broken with the test environment

File: tests/genai/judges/optimizers/test_dspy_utils.py [hunk 13]
```diff
@@ -469,3 +540,263 @@ class NonDictDemo:
...
+# Multi-assessment tests: no conflict (all assessments agree)
+@pytest.mark.parametrize(
+    ("num_assessments", "expected_label"),
+    [
+        (1, "pass"),  # Single assessment returns single-item list
+        (2, "pass"),  # Two agreeing assessments return both
+        (3, "fail"),  # Three agreeing assessments return all
+    ],
+)
+def test_trace_to_dspy_example_multi_assessment_no_conflict(mock_judge, num_assessments, expected_label):
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/test_dspy_utils.py]: can we just combine this test with the tests below and pass a list of assessments instead of having separate tests? we can create a helper to create assessments, but this way it becomes really easy to read what the assessments are and the expected label. feel free to add comments explaining what each test case does.

File: tests/genai/judges/optimizers/test_dspy_utils.py [hunk 14]
```diff
@@ -469,3 +540,263 @@ class NonDictDemo:
...
+    results = trace_to_dspy_example(trace, mock_judge)
+
+    assert len(results) == 2
+    for result in results:
+        assert result["result"] == "fail"
+
+
+# Optimizer integration test for multi-assessment handling
+def test_memalign_optimizer_handles_multi_assessment_traces(mock_judge):
+    """MemAlignOptimizer.align() should handle multiple assessments per trace."""
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/test_dspy_utils.py]: same as above, let's 
1. remove one-line docstrings
2. remove the import skip
3. move imports to the top of the file

**Review by smoorjani — CHANGES_REQUESTED** [review]

File: mlflow/genai/judges/optimizers/dspy_utils.py (hunk 4 above)
Comment (veronicalyu320) [reply mlflow/genai/judges/optimizers/dspy_utils.py]: makes sense. assessment id is passed as Pallavi hopes to see exactly which assessments are discarded, s.t. they can manually resolve conflicts in the interim.

File: tests/genai/judges/optimizers/test_dspy_utils.py [hunk 15]
```diff
@@ -469,3 +527,174 @@ class NonDictDemo:
...
+        pytest.param("empty", lambda t: [], 0, id="empty"),
+        pytest.param(
+            "wrong_judge",
+            lambda t: [_create_human_assessment("other_judge", "pass", "Wrong", t, "user")],
+            0,
+            id="wrong_judge"
+        ),
+    ],
+)
+def test_trace_to_dspy_example_edge_cases(mock_judge, scenario, assessments_fn, expected_count):
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/test_dspy_utils.py]: I'm a bit confused what we're testing here - is it that the source ID is used to filter out assessments? If so, can we create an explicit test for that instead of a generic edge case test?

File: tests/genai/judges/optimizers/test_dspy_utils.py [hunk 16]
```diff
@@ -469,3 +527,174 @@ class NonDictDemo:
...
+    captured = capsys.readouterr()
+    assert "discarded" in captured.err.lower()
+    assert "test_trace_123" in captured.err
+@pytest.mark.parametrize(
+    ("scenario", "assessments_fn", "expected_count"),
+    [
+        pytest.param(
+            "no_timestamps",
+            lambda t: [
+                Feedback(name="mock_judge", value="pass", rationale="R1",
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/test_dspy_utils.py]: nit: doesn't feel like we need to pass in a function here (harder to read) - can we just rely on _create_human_assessment and a mock time?

File: tests/genai/judges/optimizers/test_dspy_utils.py [hunk 17]
```diff
@@ -469,3 +527,174 @@ class NonDictDemo:
...
+    trace = _create_trace_with_assessments("test_mixed", assessments)
+    results = trace_to_dspy_example(trace, mock_judge)
+
+    assert len(results) == 2
+    for result in results:
+        assert result["result"] == "fail"
+
+
+def test_memalign_optimizer_handles_multi_assessment_traces(mock_judge):
+    from mlflow.genai.judges.optimizers.memalign.optimizer import MemAlignOptimizer
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/test_dspy_utils.py]: nit: let's move imports to the top of the file

File: tests/genai/judges/optimizers/test_dspy_utils.py [hunk 18]
```diff
@@ -469,3 +527,174 @@ class NonDictDemo:
...
     with pytest.raises(MlflowException, match="Demo at index 1 cannot be converted to dict"):
         format_demos_as_examples(demos, mock_judge)
+
+
+@pytest.mark.parametrize(
+    ("assessments_spec", "expected_count", "expected_label"),
+    [
+        # No conflict: all assessments agree
+        pytest.param(
+            [("pass", 1000), ("pass", 2000)],
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/test_dspy_utils.py]: can we add a comment explaining what this tuple is?

**Review by smoorjani — CHANGES_REQUESTED** [review]
logic looks good! left a few comments on the tests to address + the linter will probably have some fixes to make

**Review by smoorjani — APPROVED** [review]

**Comment by veronicalyu320** [comment]
The `windows (2)` failure on the latest run is from `tests/tracing/test_fluent.py::test_search_traces_yields_expected_dataframe_contents`, which is a known flake on the Windows runner — unrelated to this PR.

Evidence:
- Master itself hit this exact test on run [25038907580](https://github.com/mlflow/mlflow/actions/runs/25038907580) (2026-04-28T07:05 UTC), with a different assertion (`df.iloc[idx].trace_metadata == ...`).
- The previous CI run on this branch failed `windows (2)` on a different test (`tests/metrics/test_metric_definitions.py::test_bleu`, GitHub HTTP 502).
- Today's failure is on the `df.iloc[idx].spans == ...` assertion: two sibling child spans both report `start_time_unix_nano=1777403301309318100` (one is zero-duration), so `mlflow.search_traces` and `mlflow.get_trace` return them in different orders depending on Windows clock-resolution tiebreaks.
- #22783 stabilized one assertion in the same test (execution_duration ±1ms); the `spans` and `trace_metadata` assertions remain exposed to the same root cause.

Could a maintainer re-run the failed `windows (2)` job? Happy to also open a follow-up making the `spans` assertion order-insensitive if useful.

**Review by serena-ruan — APPROVED** [review]
